File: src/api/queries/n_nearest/NNearestQuery.cs
using System;
using System.Runtime.InteropServices;
using Microsoft.AspNetCore.Http;
using System.Collections;
using System.Collections.Generic;
using System.Linq;
using Microsoft.AspNetCore.Mvc;
using Microsoft.Extensions.Logging;
using DVAN.Routing;
using System.Threading.Tasks;
using DVAN.Accessibility;
using DVAN.Population;
using NetTopologySuite.Geometries;

namespace DVAN.API
{
    public class NNearestQuery
    {
        public static float[] computeQuery(double[] values, IKNNTable table, string computed_type, int count)
        {
            var results = new float[values.Length];
            for (int i = 0; i < values.Length; i++) {
                int index = i;
                List<double> vals = Enumerable.Range(0, count).Select(item => {
                    var (key, _) = table.getKNearest(index, item);
                    if (key == -1) {
                        return -9999;
                    }
                    return values[key];
                }).Where(item => item != -9999).ToList();
                if (computed_type == "min") {
                    if (vals.Count == 0) {
                        results[i] = -9999;
                        continue;
                    }
                    results[i] = (float)vals.Min();
                }
                if (computed_type == "max") {
                    if (vals.Count < count) {
                        results[i] = -9999;
                        continue;
                    }
                    results[i] = (float)vals.Max();
                }
                if (computed_type == "median") {
                    if (vals.Count < count) {
                        results[i] = -9999;
                        continue;
                    }
                    vals.Sort();
                    if (vals.Count % 2 == 1) {
                        var key = (vals.Count - 1) / 2;
                        results[i] = (float)vals[key];
                    }
                    else {
                        var key1 = (vals.Count - 2) / 2;
                        var key2 = (vals.Count - 2) / 2 + 1;
                        results[i] = (float)(vals[key1] + vals[key2]) / 2;
                    }
                }
                if (computed_type == "mean") {
                    if (vals.Count < count) {
                        results[i] = -9999;
                        continue;
                    }
                    else {
                        results[i] = (float)(vals.Sum() / vals.Count);
                    }
                }
                if (computed_type == "sum") {
                    if (vals.Count < count) {
                        results[i] = -9999;
                        continue;
                    }
                    else {
                        results[i] = (float)vals.Sum();
                    }
                }
            }

            return results;
        }
    }
}

File: src/main/java/org/tud/oas/api/queries/n_nearest/NNearestQuery.java
package org.tud.oas.api.queries.n_nearest;

import java.util.ArrayList;
import java.util.List;

import org.tud.oas.routing.IKNNTable;
import org.tud.oas.supply.ISupplyView;

class NNearestQuery {

    public static float[] computeQuery(ISupplyView supply, IKNNTable table, String computedType, int count) {
        float[] results = new float[supply.pointCount()];
        for (int i = 0; i < supply.pointCount(); i++) {
            int index = i;
            List<Double> vals = new ArrayList<>();
            for (int item = 0; item < count; item++) {
                int key = table.getKNearest(index, item);
                if (key == -1) {
                    vals.add(-9999.0);
                } else {
                    vals.add((double) supply.getSupply(key));
                }
            }
            vals.removeIf(item -> item == -9999.0);

            if (computedType.equals("min")) {
                if (vals.isEmpty()) {
                    results[i] = -9999;
                } else {
                    results[i] = (float) vals.stream().mapToDouble(Double::doubleValue).min().orElse(-9999.0);
                }
            } else if (computedType.equals("max")) {
                if (vals.size() < count) {
                    results[i] = -9999;
                } else {
                    results[i] = (float) vals.stream().mapToDouble(Double::doubleValue).max().orElse(-9999.0);
                }
            } else if (computedType.equals("median")) {
                if (vals.size() < count) {
                    results[i] = -9999;
                } else {
                    vals.sort(Double::compareTo);
                    if (vals.size() % 2 == 1) {
                        int key = (vals.size() - 1) / 2;
                        results[i] = vals.get(key).floatValue();
                    } else {
                        int key1 = (vals.size() - 2) / 2;
                        int key2 = (vals.size() - 2) / 2 + 1;
                        results[i] = (float) ((vals.get(key1) + vals.get(key2)) / 2);
                    }
                }
            } else if (computedType.equals("mean")) {
                if (vals.size() < count) {
                    results[i] = -9999;
                } else {
                    results[i] = (float) (vals.stream().mapToDouble(Double::doubleValue).sum() / vals.size());
                }
            } else if (computedType.equals("sum")) {
                if (vals.size() < count) {
                    results[i] = -9999;
                } else {
                    results[i] = (float) vals.stream().mapToDouble(Double::doubleValue).sum();
                }
            }
        }
        return results;
    }
}
